### DISCUSSION TITLE
Randomly changing results from predictions

### DISCUSSION THREAD

**Discussion by sanmayshelat** [moj-analytical-services/splink#2530] [Q&A] [answered]
Hi,

I have a model saved that seems to (generally) have parameters that are expected/explainable. I am loading this model with `linker = Linker(df, settings="model_name.json", db_api=db_api)` and applying `df_predict = linker.inference.predict(threshold_match_weight=2,).as_pandas_dataframe()`. However, each time I reload the kernel I am getting a slightly different length for df_predict. This doesn't happen when I re-run `linker = Linker(df...` or even if I create a separate version like `linker2 = Linker(df...`. Any suggestions on why this might be happening?

I'm using Splink=4.0.5 and Python=3.12.3.

**Comment by ADBond** [comment, marked as answer]
Hmmm, that is puzzling. Nothing immediately leaps to mind as the likely cause - if you are not retraining the model then `predict()` should give the same set of results consistently for the same input data.

Any further details you are able to provide (the `.json` file, or your comparisons, or the script you are running, backend) might point to some clues. Are there any possibly non-deterministic functions in your comparisons (or blocking rules)? And is the input data definitely unchanged between these runs?

One thing that might help narrow it down is to try and investigate the record pairs that are not always appearing - if you save the left+right ids of your `df_predict` table to file you can compare across runs and then see if there is anything noteworthy about them (and perhaps share any details you are able to here).

**Comment by sanmayshelat** [comment]
Thanks for getting back to me so quickly! It appears that the randomness is coming from a data manipulation step in Polars. I was able to get repeatable results using Pandas end to end or using a saved, post-manipulation dataset. I will look into what it is in Polars that's causing the randomness.

**Reply by RobinL** [reply]
Thanks for the update, glad you resolved the problem
